### ISSUE TITLE
How to debug the object IBusPinyinEngine's startup procedure

### ISSUE THREAD

**Issue by BSoD-Ultimate** [libpinyin/ibus-libpinyin#171]
#### Which distribution and version?: 
openSUSE Leap 15.0 Linux localhost 4.12.14-lp150.12.48-default x86_64

#### Which desktop environment and version?: 
GNOME 3.26.2

#### Which session type?: 
Wayland

#### IBus version?: 
IBus 1.5.17 & ibus-libpinyin 1.9.1

#### Issue description:
The company I worked is working on porting a Chinese pinyin input method running on Windows to Linux. Now we have to port the engine to the IBus platform.

To understand how the IBus input engine works and how to implement an input method based on the IBus platform correctly, I downloaded Then I successfully compiled and installed an ibus-libpinyin build with no optimization on the Linux system.

The debug build of ibus-libpinyin runs with no problems. I could use Qt Creator's debugger front-end to toggle breakpoints on most of the input method's source code. **When Attempting to set breakpoints on the function `ibus_pinyin_engine_class_init` locates in `PYEngine.cc`, I can't find out in which condition the code will be executed.** 

By extracting the precompiled RPM I've already known that to register an input method on IBus engine, the developer must write an XML locate in the IBus components folder. e.g. the folder `/usr/share/ibus/component`. I attempted to debug the input method after adding ibus-libpinyin as an available input source, while the current input method is not switched to ibus-libpinyin. Then I found out the process will not be started until I switch the current input method to it. The attempt fails.

I also noticed that in the file `PYMain.cc`, the code would register a debug version of the input engine programmatically if the input engine executable is started without the `--ibus` argument. So I made another attempt. I started an input method process without any startup arguments when ibus-daemon is running, expecting the process will register the engine by itself. Unfortunately, after these actions, I did not see any 'debug' version of ibus-libpinyin listed in the `IBus Preferences` dialog. There is also a pull request #95 mentioning that the debug version's registration of input engines does not work.

As both of the attempts failed I want to know how to debug the input engine object's startup procedure. We may also encounter this problem when porting our input engine to the IBus platform.

Sorry for I don't know much about GObject's API as the IBus platform heavily uses it.

**Comment by epico** [comment]
Which Chinese input method are you porting?

Usually I just use gdb to debug with the command line `./ibus-engine-libpinyin --ibus`.

**Comment by BSoD-Ultimate** [comment]
Thank you. Later I tried starting the ibus-libpinyin engine via Qt Creator with the command line `./ibus-engine-libpinyin --ibus`, Then I killed the input method process created by the IBus input engine. After these, I'm able to break the execution on the code I previously said.

By the way, the `libpinyin-debug` programmatically registered in `PYMain.cc` really seems takes no effect and might be ignored. I also tried the Tumbleweed version of openSUSE then the 'debug' version of the input method did not appear on the preferences dialog either.

The Chinese input method we are porting is unispim.
